theano/tensor/nnet/ctc.py

```diff
@@ -0,0 +1,113 @@
+
+    def __init__(self, compute_grad=True):
+        if not compute_grad:
+            self.func_name = "APPLY_SPECIFIC(ctc_cost_cpu_no_grad)"
+
+        gof.COp.__init__(self, self.func_file, self.func_name)
+        gof.OpenMPOp.__init__(self)
+
+        self.compute_grad = compute_grad
+        self.costs = T.fvector(name="ctc_cost")
+        if self.compute_grad:
+            self.gradients = T.ftensor3(name="ctc_grad")
+
+        if config.ctc.root == "":
+            raise ValueError('ctc.root variable is not set, please set it '
+                             'to the root directory of the CTC library in '
+                             'your system.')
+
+    def c_lib_dirs(self):
+        dirs = []
+        if ctc_enabled:
+            # We assume here that the compiled library (libwarpctc.so) is available
+            # at the build directory of the CTC root directory.
+            dirs.append(os.path.join(config.ctc.root, "build"))
+        return dirs
+
+    def c_libraries(self):
+        return ["warpctc"]
+
+    def c_header_dirs(self):
+        dirs = []
+        if ctc_enabled:
+            # We assume here that the header is available at the include directory
+            # of the CTC root directory.
+            dirs.append(os.path.join(config.ctc.root, "include"))
+        return dirs
+
+    def c_compile_args(self):
+        return gof.OpenMPOp.c_compile_args(self)
+
+    def c_headers(self):
+        return ["ctc.h"] + gof.OpenMPOp.c_headers(self)
+
+    def make_node(self, activations, labels, input_lengths):
+        if not ctc_enabled:
+            raise RuntimeError('Baidu CTC is not enabled and '
+                               'ConnectionistTemporalClassification Op '
+                               'can not be constructed.')
+        t_activations = T.as_tensor_variable(activations)
+        # Ensure activations array is C-contiguous
+        t_activations = cpu_contiguous(t_activations)
+
+        t_labels = T.as_tensor_variable(labels)
+        t_input_lengths = T.as_tensor_variable(input_lengths)
+
+        if t_activations.type.dtype != 'float32':
+            raise TypeError('Activations must use the float32 type!')
+
+        if t_labels.type.dtype != 'int32':
+            raise TypeError('Labels must use the int32 type!')
+
+        if t_input_lengths.type.dtype != 'int32':
+            raise TypeError('Label lengths must use the int32 type!')
+
+        # Return only the cost. Gradient will be returned by grad()
+        self.default_output = 0
+
+        return gof.Apply(self, inputs=[t_activations, t_labels, t_input_lengths],
+                         outputs=[self.costs, self.gradients])
+
+    def grad(self, inputs, output_grads):
+        if not ctc_enabled:
+            raise RuntimeError('Baidu CTC is not enabled and '
+                               'ConnectionistTemporalClassification Op '
+                               'can not be constructed.')
+        return [self.gradients,
```

Comment:
In fact, `grad()` method in Theano is not intended to simply return the gradient of the Op's output with respect to the Op's inputs. The full documentation can be found here: http://deeplearning.net/software/theano_versions/dev/extending/op.html#gradient . I'll try to explain that in following lines:

Theano works on graphs of Ops and want to be able to compute the gradient of any output of full graph with respect to any initial input of the graph.

So, for example, let's take a simple Op that performs a function `f` so that `f(x) = y` (`x` in the Op's input, `y` is the Op's output). Then, Op's `grad()` method will receive the gradient of the graph output `C` (graph in which current Op is present) with respect to the Op's output `y`: that is, `dC/dy`. The goal of `grad()` method is to return the gradient of graph output `C` with respect to the Op's input `x`: that is, `dC/dx`. To do that, `grad()` method generally compute `dy/dx` (the real classic gradient of Op's output with respect to Op's input), and then computes `dC/dx` by multiplying `dC/dy` with `dy/dx`:
```
dC/dx = dC/dy * dy/dx
```
`dC/dy` is in `output_grads` parameters. In practice, `output_grads` contains one `dC/dy` for each Op's output `y`. You should then return `dC/dx` for each Op's input `x`.

Here, for first input (activations), `dy/dx` is indeed `self.gradients`, and `dC/dy` is `output_grads[0]` (first output is the cost), then we should "multiply" these two variables. As CTC works like a batched Op (multiple independent activations sequences), we must be careful about how we perform multiplication. I found the following code to do that:
```python
        import theano
        z = output_grads[0]
        g = theano.tensor.basic.batched_dot(z, self.gradients.dimshuffle(1, 0, 2)).dimshuffle(1, 0, 2)
```
(each cost is relative to each sequence of activations, but activations sequences are in the second dimension of activations tensor (minibatch size), so the first call to `dimshuffle()` puts that second dimension of activations to the front, so that each gradient wrt/ cost multiplies its corresponding sequence of gradients wrt activations; at the end, second call to `dimshuffle()` put the tensor back to its original dimensions order).

Then, you could return `[g, grad_undefined..., grad_undefined...]` instead of `[self.gradients, grad_undefined..., grad_undefined...]` (gradient is still undefined for other inputs, ie. labels and input lengths).